include/revng/Yield/Support/PTML.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <algorithm>
+#include <climits>
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringMap.h"
+
+#include "revng/ADT/Concepts.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/YAMLTraits.h"
+
+namespace yield::ptml {
+
+namespace tags {
+
+inline constexpr auto Div = "div";
+inline constexpr auto Span = "span";
+
+} // namespace tags
+
+namespace attributes {
+
+inline constexpr auto scope = "data-scope";
+inline constexpr auto token = "data-token";
+inline constexpr auto locationDefinition = "data-location-definition";
+inline constexpr auto locationReferences = "data-location-references";
+inline constexpr auto modelEditPath = "data-model-edit-path";
+
+} // namespace attributes
+
+namespace scopes {
+
+inline constexpr auto Annotation = "annotation";
+inline constexpr auto Comment = "comment";
+inline constexpr auto Error = "error";
+inline constexpr auto Link = "link";
+
+} // namespace scopes
+
+namespace tokens {
+
+inline constexpr auto Indentation = "indentation";
+inline constexpr auto Comment = "comment";
+
+} // namespace tokens
+
+class Tag {
+private:
+  std::string TheTag;
+  std::string Content;
+  llvm::StringMap<std::string> Attributes;
+
+public:
+  Tag() {}
+  explicit Tag(llvm::StringRef Tag) : TheTag(Tag.str()) {}
+  explicit Tag(llvm::StringRef Tag, llvm::StringRef Content) :
+    TheTag(Tag.str()), Content(Content.str()) {}
+
+  Tag &setContent(const llvm::StringRef Content) {
+    this->Content = Content.str();
+    return *this;
+  }
+
+  Tag &add(llvm::StringRef Name, llvm::StringRef Value) {
+    Attributes[Name] = Value.str();
+    return *this;
+  }
+
+  template<range_with_value_type<llvm::StringRef> T>
+  Tag &add(llvm::StringRef Name, const T &Values) {
+    Attributes[Name] = llvm::join(Values, ",");
+    return *this;
+  }
+
+  template<typename... T>
+  requires(std::is_convertible_v<llvm::StringRef, T> and...)
+    Tag &add(llvm::StringRef Name, T &&...Values) {
+    Attributes[Name] = llvm::join(",", std::forward<T>(Values)...);
+    return *this;
+  }
+
+  std::string open() const {
+    llvm::SmallString<128> Out;
+    Out.append({ "<", TheTag });
+    for (auto &Pair : Attributes)
+      Out.append({ " ", Pair.first(), "=\"", Pair.second, "\"" });
+    //Out.pop_back();
+    Out.append({'>'});
+    return Out.str().str();
+  }
+
+  std::string close() const { return "</" + TheTag + ">"; }
+
+  std::string serialize() const { return open() + Content + close(); }
```

Comment:
Have you considered a RAII approach for this?

Reply:
Can you paste an example output? I cannot appreciate the nuances otherwise.
All of this seems to have a lot of non-obvious implicit rules which would benefit a lot from being wrapped in a RAII object and we should really really strive for it.

Reply:
Ok, I was mistaken, RAII works really well (dunno what happened the first time I tried it, probably did something wrong)
Added to PR

Reply:
Great!